contrib/completion/_fd

```diff
@@ -0,0 +1,210 @@
+  fd_args=(
+    + '(hidden)' # hidden files
+    {-H,--hidden}'[search hidden files/directories]'
+
+    + '(no-ignore-full)' # all ignore files
+    '(no-ignore-partial)'{-I,--no-ignore}"[don't respect .(git|fd)ignore and global ignore files]"
+    $no'(no-ignore-partial)*'{-u,--unrestricted}'[alias for --no-ignore, when repeated also alias for --hidden]'
+
+    + no-ignore-partial # some ignore files
+    "(no-ignore-full --no-ignore-vcs)--no-ignore-vcs[don't respect .gitignore files]"
+    "(no-ignore-full --no-global-ignore-file)--no-global-ignore-file[don't respect the global ignore file]"
+
+    + '(case)' # case-sensitivity
+    {-s,--case-sensitive}'[perform a case-sensitve search]'
+    {-i,--ignore-case}'[perform a case-insensitve search]'
+
+    + '(regex-pattern)' # regex-based search pattern
+    '(no-regex-pattern)--regex[perform a regex-based search (default)]'
+
+    + '(no-regex-pattern)' # non-regex-based search pattern
+    {-g,--glob}'[perform a glob-based search]'
+    {-F,--fixed-strings}'[treat pattern as literal string instead of a regex]'
+
+    + '(match-full)' # match against full path
+    {-p,--full-path}'[match the pattern against the full path instead of the basename]'
+
+    + '(follow)' # follow symlinks
+    {-L,--follow}'[follow symbolic links to directories]'
+
+    + '(abs-path)' # show absolute paths
+    '(long-listing)'{-a,--absolute-path}'[show absolute paths instead of relative paths]'
+
+    + '(null-sep)' # use null separator
+    '(long-listing)'{-0,--print0}'[separate search results by the null character]'
+
+    + '(long-listing)' # long-listing output
+    '(abs-path null-sep max-results exec-cmds)'{-l,--list-details}'[use a long listing format with file metadata]'
+
+    + '(max-results)' # max number of results
+    '(long-listing exec-cmds)--max-results=[limit number of search results to given count and quit]:count'
+    '(long-listing exec-cmds)-1[limit to a single search result and quit]'
+
+    + '(error-messages)' # show FS errors
+    $no'--show-errors'
+
+    + '(file-system)' #
+    "--one-file-system[don't descend into directories on other file systems]"
+    '!--mount'
+    '!--xdev'
+
+    + dir-depth # directory depth
+    '(--exact-depth -d --max-depth)'{-d+,--max-depth=}'[set max directory depth to descend when searching]:depth'
+    '!(--exact-depth -d --max-depth)--maxdepth:depth'
+    '(--exact-depth --min-depth)--min-depth=[set directory depth to descend before start searching]:depth'
+    '(--exact-depth -d --max-depth --maxdepth --min-depth)--exact-depth=[only search at the exact given directory depth]:depth'
+
+    + filter-misc # Filter search
+    '*'{-t+,--type=}"[filter search by type]:type:(($fd_file_types))"
+    '*'{-e+,--extension=}'[filter search by file extension]:extension'
+    '*'{-E+,--exclude=}'[exclude files/directories that match the given glob pattern]:pattern'
+    '*'{-S+,--size=}'[limit search by file size]:size limit'
```

Comment:
For arguments like this it's common to describe the units/format in parentheses after the description. So this one might be like `size limit (bytes)`. If it accepts stuff like `100k` and `1G`, i'm not sure there's a common way of wording that, but maybe something like `size limits (bytes with optional suffix)`

Reply:
I followed the man page and use now `(format: <+-><num><unit>)`.

Reply:
I added completion for `--size` now, so I remove the `(format ...)` from the option description and added it to the completion message.

``` zsh
if compset -P '[-+][0-9]##'; then
  local -a suff=(
    'B:bytes'
    'K:kilobytes  (10^3  = 1000   bytes)'
    'M:megabytes  (10^6  = 1000^2 bytes)'
    'G:gigabytes  (10^9  = 1000^3 bytes)'
    'T:terabytes  (10^12 = 1000^4 bytes)'
    'Ki:kibibytes  ( 2^10 = 1024   bytes)'
    'Mi:mebibytes  ( 2^20 = 1024^2 bytes)'
    'Gi:gigibytes  ( 2^30 = 1024^3 bytes)'
    'Ti:tebibytes  ( 2^40 = 1024^4 bytes)'
  )
  _describe -t units 'size limit units' suff -V 'units'
elif compset -P '[-+]'; then
  _message -e 'size limit number (full format: <+-><number><unit>)'
else
  _values 'size limit prefix (full format: <prefix><number><unit>)' \
    '\+[file size must be greater or equal to]'\
    '-[file size must be less than or equal to]' && ret=0
fi
```

It first completes the prefix (`+` or `-`), and then after a number is inserted it completes the units.
If the prefix was typed manually and then <kbd>Tab</kbd> is pressed, it will show in the completion message that a number is needed.

@sharkdp the convention for GNU utils (find, du, ls, etc) is to use K,M,G,T... for powers of 1024, and KB,MB,GB,TB... for powers of 1000. Binary prefixes are also supported, with KiB=K, MiB=M, etc.

But fd is using K,M,G,T for powers of 1000. I think it may be good to follow convention.